### ISSUE TITLE
clang trunk missed the buffer-overflow in code

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#55189]
For the following code, both `clang-tk -fsanitize=address -O0` and `-O3` do not report the global-buffer-overflow of `a[]`. Interestingly, if you change `M` to a smaller value (e.g., `#define M 2`), they could detect it. Initially I thought it may relate to optimizations as discussed in https://github.com/llvm/llvm-project/issues/53972. But it's clearly not because changing `#define M 5` affected the result and even `-O0` missed the bug.
```c
#define M 5
volatile int a[M];
__attribute__((__noinline__))
void b(x) {
    if(x) a[M] = 7;
}
main(){ 
    b(M); 
    __builtin_printf("a[M]=%d\n", a[M]);
}
```

**Comment by efriedma-quic** [comment]
I think the significance of "M" is related to alignment.  The x86-64 ABI requires that global arrays of at least 16 bytes are 16-byte aligned... and the accesses that aren't getting diagnosed fall into that "padding" (not actually padding, but apparently asan thinks it is).

**Comment by shao-hua-li** [comment]
So, is there any workaround to this problem? Or could I consider this as an ASAN issue/bug?

**Comment by efriedma-quic** [comment]
It's a bug.  I don't know of any workaround.
